presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -2031,55 +2051,100 @@ public void run()
         {
-            return format("add table %s.%s", table.getDatabaseName(), table.getTableName());
+            return format("add table %s.%s", newTable.getDatabaseName(), newTable.getTableName());
         }
 
         public void run(ExtendedHiveMetastore metastore)
         {
             try {
-                metastore.createTable(table, privileges);
+                metastore.createTable(newTable, privileges);
                 done = true;
             }
             catch (RuntimeException e) {
-                try {
-                    Optional<Table> remoteTable = metastore.getTable(table.getDatabaseName(), table.getTableName());
-                    // getPrestoQueryId(partition) is guaranteed to be non-empty. It is asserted in the constructor.
-                    if (remoteTable.isPresent() && getPrestoQueryId(remoteTable.get()).equals(getPrestoQueryId(table))) {
+                handleException(metastore, e);
+            }
+        }
+
+        private void handleException(ExtendedHiveMetastore metastore, RuntimeException exception)
+        {
+            try {
+                Optional<Table> existingTable = metastore.getTable(newTable.getDatabaseName(), newTable.getTableName());
+                existingTable.ifPresent(table -> {
+                    // getPrestoQueryId(newTable) is guaranteed to be non-empty. It is asserted in the constructor.
+                    if (getPrestoQueryId(table).equals(getPrestoQueryId(newTable))) {
                         done = true;
                     }
-                }
-                catch (RuntimeException ignored) {
-                    // When table could not be fetched from metastore, it is not known whether the table was added.
-                    // Deleting the table when aborting commit has the risk of deleting table not added in this transaction.
-                    // Not deleting the table may leave garbage behind. The former is much more dangerous than the latter.
-                    // Therefore, the table is not considered added.
-                }
-                if (!done) {
-                    throw e;
+                    else {
+                        // if the table definition in the metastore is different than what this tx wants to create
+                        // then there is a conflict, e.g., current tx wants to create T(a: bigint),
+                        // but another tx already created T(a: varchar)
+                        // this may be a problem if there is an insert after this step, so we fail early.
+                        if (!hasTheSameSchema(newTable, table)) {
+                            throw new TableAlreadyExistsException(
+                                    new SchemaTableName(newTable.getDatabaseName(), newTable.getTableName()),
+                                    format("Table already exists with a different schema: '%s'", newTable.getTableName()),
+                                    exception);
```

Comment:
I would inline `handleException` and change this line to `e = new TableAlreadyExistsException(...`